app/python_apps/aiohttp_ratelimit.py

```diff
@@ -0,0 +1,103 @@
+# A demo case of async request get with rate limit.
+# pip install aiohttp aiohttp-client-manager bucketratelimiter tqdm
+from typing import Tuple, Any, Union, List, Coroutine
+
+import asyncio
+from asyncio import events, coroutines
+from datetime import datetime
+from asyncio import tasks
+
+from bucketratelimiter import AsyncioBucketTimeRateLimiter
+import aiohttp_client
+from tqdm.asyncio import tqdm
+
+
+async def async_tasks(cors: List[Coroutine]) -> List:
+    results = []
+    with tqdm(total=len(cors)) as pbar:
+        for next_to_complete in asyncio.as_completed(cors):
+            answer = await next_to_complete
+            results.append(answer)
+            pbar.update()
+    return results
+
+
+def cancel_all_async_tasks(loop):
+    to_cancel = tasks.all_tasks(loop)
+    if not to_cancel:
+        return
+
+    for task in to_cancel:
+        task.cancel()
+
+    loop.run_until_complete(tasks.gather(*to_cancel, loop=loop, return_exceptions=True))
+
+    for task in to_cancel:
+        if task.cancelled():
+            continue
+        if task.exception() is not None:
+            loop.call_exception_handler(
+                {
+                    "message": "unhandled exception during asyncio.run() shutdown",
+                    "exception": task.exception(),
+                    "task": task,
+                }
+            )
+
+
+class AsyncApi:
+    limiter = AsyncioBucketTimeRateLimiter(max_size=10, recovery_time=5.0, rest_time=0.5)
+
+    def __init__(self):
+        self._loop = asyncio.new_event_loop()
+        events.set_event_loop(self.loop)
+        self.limiter.activate()
+
+    def __del__(self):
+        try:
+            self.loop.run_until_complete(self.loop.shutdown_asyncgens())
+        finally:
+            cancel_all_async_tasks(self._loop)
+            self.limiter.deactivate()
+            events.set_event_loop(None)
+            self.loop.close()
+
+    def run_until_complete(self, run_func, *, debug=None):
+        if not coroutines.iscoroutine(run_func):
+            raise ValueError("a coroutine was expected, got {!r}".format(run_func))
+        if debug is not None:
+            self.loop.set_debug(debug)
+        return self.loop.run_until_complete(run_func)
+
+    @property
+    def loop(self):
+        return self._loop
+
+    @limiter
+    async def get_data(self, url: str, name: str):
+        print(f"{datetime.now()} - Start get_data {name}...")
+        async with aiohttp_client.get(url) as response:
+            html = await response.text()
+        print(f"{datetime.now()} - Done get_data {name} - {html[:15]}...")
+
+
+def main():
+    api = AsyncApi()
+    task_num = 20
+    cors = []
+    url = "http://python.org"
+    print("start first!")
+    for i in range(task_num):
```

Comment:
It is good enough for a cheatsheet.

I prefer to implement it as a context manager if it serves as a user interface.

Then the following code
```
    cors = []
    for i in range(task_num, task_num + task_num):
        cor = api.get_data(url, f"task_{i}")
        cors.append(cor)
    api.run_until_complete(async_tasks(cors))
````
could be replaced with code like

```
with AsyncAPI() as f:
      f.append(get_data, url, f"task_{i}")
```
`AsyncioBucketTimeRateLimiter` will be created in the `__enter__` step.
`get_data` will be decorated in `f.append`  and the corresponding coroutine will be created.
`run_until_complete` will called in the `__exit__` step.

With this design,  the user interface will be easier.
Besides, `AsyncApi`  will have a re-usable definition, i.e. creating a container for normal function calls and then run them with `asyncio` under customized limitation.

Reply:
I don't implement this in context manager because I have another class that uses the async API as a class member. That class will be a singleton class and will be deleted when the executive is terminated. So I guess I don't need an API as a context manager. I am unsure if I am right or wrong in the above case. Feel free to give me some feedback.

But I think the async API can be designed to support context manager and original mode by adding `__enter__ ` and `__exit__` like `AsyncioBucketTimeRateLimiter` do. So people can choose what they need in their scenario. 
Since I am not familiar with the context manager, I prefer to merge this first. You might implement a context manager version. I can review it.

By the way, Does neovim support grammar/spell checks? I use Pycharm and it supports powerful grammar and spell check.

Reply:
Thanks. I'll merge it first.

I use this "coc-zi" for spell-checking before (grammar checking is not supported)
https://github.com/you-n-g/deploy/blob/master/configs/nvim/yx/plugs/coc_conf.vim#L187

But I switch to neovim native lsp now. I didn't set a spell checker for me.